
# Week 01 Assignment  
## Data Quality, Evaluation, Scaling, and Encoding

**Student name: Farhan Bin Rabbani**   

This is a small assignment that connects topics from Module 1, 2, and 3.  
You must complete it in this Colab notebook.

You will need to use concepts that appeared in the videos:
- Module 1 and 2: basic descriptive statistics, proportions, confusion matrix, accuracy, precision, recall
- Module 3: standardization, min max scaling, nominal vs ordinal, one hot encoding, ordinal encoding, Euclidean and Manhattan distance

Please do not use any extra libraries beyond `pandas`, `numpy`.



---
## 0. Setup and Dataset

We will use a dataset that should have columns given below:

- `user_id`  
- `age`  
- `monthly_income` (numeric)  
- `daily_screen_time_min` (numeric)  
- `daily_app_opens` (numeric)  
- `true_label` and `pred_label` for a binary classification task (0 or 1)  
- `satisfaction_level` (for example: `Low`, `Medium`, `High`)  
- `city_type` (for example: `Urban`, `Suburban`, `Rural`)


In [41]:
# Cell 1: Imports
import pandas as pd
import numpy as np

In [42]:
# Cell 2: Load the dataset (Already done for you)
df = pd.read_csv("https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220")

# Show first few rows
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  
0           0           0             Medium  Suburban  
1           0           0                Low     Urban  
2           1           0               High     Rural  
3           1           0               High  Suburban  
4           0           0             Medium  Suburban


### 0.1 Check your dataset

1. Confirm that the dataset loaded correctly.  
2. Check that you have at least these columns:  
   - numeric: `age`, `monthly_income`, `daily_screen_time_min`, `daily_app_opens`  
   - labels: `true_label`, `pred_label`  
   - categorical: `satisfaction_level`, `city_type`  



---
## Part A - Module 1 and 2 Review

In this part you will do simple descriptive statistics and basic classification evaluation.



### Q1. Descriptive statistics on a numeric feature

Choose one numeric column, for example `daily_screen_time_min`.


In [43]:
# Q1.1: Choose your numeric column here [We already write this ans]
num_col = "daily_screen_time_min"

df[num_col].describe()

count    100.000000
mean     181.890000
std       68.886951
min       60.000000
25%      122.000000
50%      178.000000
75%      243.750000
max      299.000000
Name: daily_screen_time_min, dtype: float64


> **Q1.2 Short answer: [Marks: 05]**  
> Look at the count, mean, min, max, and standard deviation for your chosen column.  
> In 2 to 3 sentences, comment on what you see.  
> For example, does the max look very far from the mean, or does it look quite close?

Write your answer here:

>  count: 100; mean: 181.89; min: 60; max: 299; standard deviation: 68.887.

>  The max looks pretty far from the mean value.
>  Because the difference between the values is quite high.



### Q2. Proportion of positive class

Use the `true_label` column, where 1 means "positive" and 0 means "negative".


In [44]:
# Q2.1: Compute proportion of positive class [We already write this ans]
label_col = "true_label"

positive_count = (df[label_col] == 1).sum()
total_count = df.shape[0]
positive_proportion = positive_count / total_count

print("Positive count:", positive_count)
print("Total samples:", total_count)
print("Proportion of positive class:", positive_proportion)

Positive count: 52
Total samples: 100
Proportion of positive class: 0.52



> **Q2.2 Short answer: [5 marks]**  
> In 1 to 2 sentences, explain what this proportion tells you about your dataset.  
> For example, is the dataset balanced between 0 and 1, or is one class much more common?

Write your answer here:

>  The proportion of positive class is 0.52.
>  It means that the dataset is fairly balanced between 0 and 1, as the proportions of 0's and 1's aren't too far.



### Q3. Confusion matrix and basic metrics

For this question, use:
- `true_label` as the actual label  
- `pred_label` as the model prediction


In [45]:
# Q3.1: Manually compute TP, TN, FP, FN [We already write this ans]
true_col = "true_label"
pred_col = "pred_label"

tp = ((df[true_col] == 1) & (df[pred_col] == 1)).sum()
tn = ((df[true_col] == 0) & (df[pred_col] == 0)).sum()
fp = ((df[true_col] == 0) & (df[pred_col] == 1)).sum()
fn = ((df[true_col] == 1) & (df[pred_col] == 0)).sum()

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

TP: 28
TN: 27
FP: 21
FN: 24


In [46]:
# Q3.2: Compute accuracy, precision, recall [We already write this ans]
accuracy = (tp + tn) / (tp + tn + fp + fn) if (tp + tn + fp + fn) > 0 else 0

precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall = tp / (tp + fn) if (tp + fn) > 0 else 0

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)

Accuracy: 0.55
Precision: 0.5714285714285714
Recall: 0.5384615384615384



> **Q3.3 Short answer: [10 marks]**  
> In 3 to 4 sentences, briefly comment on the model using these three metrics.  
> For example, is the model catching most positives (high recall) or being careful when it predicts positive (high precision)?

Write your answer here:

>  The accuracy is 55%, hence it can correctly identify the labels a  little more than half the number of tries.

>  The precision is around 57%, so it's predicting the positive values a little more than half the amount of tries.

>  The recall is around 53%, meaning it is finding the actual positives a little more than half the amount of tries.



---
## Part B - Module 3: Scaling and Encoding

Now we will pick a few features and apply scaling and encoding.



### Q4. Standardization and Min max scaling

Use one numeric column, `monthly_income`.


In [47]:
# Q4.1: Choose the numeric column [2 marks]
df['monthly_income'].head()

0    3734.19
1    2594.19
2    3550.47
3    3821.18
4    1750.84
Name: monthly_income, dtype: float64

In [48]:
# Q4.2: Standardization with z-score [10 marks]
mean_val = df['monthly_income'].mean()
std_val = df['monthly_income'].std()

df['monthly_income_z_score'] = (df['monthly_income'] - mean_val) / std_val
df['monthly_income_z_score'].head()

0    0.944685
1   -0.324626
2    0.740126
3    1.041542
4   -1.263639
Name: monthly_income_z_score, dtype: float64

In [49]:
# Q4.3: Min max scaling implementation [10 marks]
min_val = df['monthly_income'].min()
max_val = df['monthly_income'].max()
rg = max_val - min_val

df['monthly_income_min_max'] = (df['monthly_income'] - min_val) / rg
df['monthly_income_min_max'].head()

0    0.675209
1    0.393685
2    0.629839
3    0.696691
4    0.185420
Name: monthly_income_min_max, dtype: float64


> **Q4.4 Short answer: [3 marks]**  
> Compare the standardized and min max scaled columns in 2 to 3 sentences.  
> Mention what kind of range each one uses and how the numbers look.

Write your answer here:

>  The standardized scaling has mean of 0 and deviation of points is not more than 1.
>  The values seem to have values in both negative and positive regions.
>  On the other hand, min-max scaling has scaled values ranging from 0 to 1.



### Q5. One hot and ordinal encoding

We will use:
- `city_type` as a nominal feature  
- `satisfaction_level` as an ordinal feature with order `Low` < `Medium` < `High`  


In [50]:
# Q5.1: One hot encoding for city_type using pandas [10 marks]
d_city = pd.get_dummies(df['city_type'], prefix = 'city', dtype = int)
df_encode = pd.concat([df, d_city], axis = 1)
df_encode = df_encode.drop('city_type', axis = 1)
df_encode.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level  monthly_income_z_score  \
0           0           0             Medium                0.944685   
1           0           0                Low               -0.324626   
2           1           0               High                0.740126   
3           1           0               High                1.041542   
4           0           0             Medium               -1.263639   

   monthly_income_min_max  city_Rural  city_Suburban  city_Urban  
0                0.675209           0              1           0  
1                0.393685           0              0           1  
2                0.629839           1              0           0  
3                0.696691           0              1           0  
4                0.185420           0              1           0

In [51]:
# Q5.2: Attach one hot encoded columns to df [5 marks]
df = pd.concat([df, d_city], axis = 1)
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  \
0           0           0             Medium  Suburban   
1           0           0                Low     Urban   
2           1           0               High     Rural   
3           1           0               High  Suburban   
4           0           0             Medium  Suburban   

   monthly_income_z_score  monthly_income_min_max  city_Rural  city_Suburban  \
0                0.944685                0.675209           0              1   
1               -0.324626                0.393685           0              0   
2                0.740126                0.629839           1              0   
3                1.041542                0.696691           0              1   
4               -1.263639                0.185420           0              1   

   city_Urban  
0           0  
1           1  
2           0  
3           0  
4           0

In [52]:
# Q5.3: Ordinal encoding for satisfaction_level [10 marks]
order = {'Low':1, 'Medium':2, 'High':3}
df['satisfaction_level'] = df['satisfaction_level'].map(order).astype(float)
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label  satisfaction_level city_type  \
0           0           0                 2.0  Suburban   
1           0           0                 1.0     Urban   
2           1           0                 3.0     Rural   
3           1           0                 3.0  Suburban   
4           0           0                 2.0  Suburban   

   monthly_income_z_score  monthly_income_min_max  city_Rural  city_Suburban  \
0                0.944685                0.675209           0              1   
1               -0.324626                0.393685           0              0   
2                0.740126                0.629839           1              0   
3                1.041542                0.696691           0              1   
4               -1.263639                0.185420           0              1   

   city_Urban  
0           0  
1           1  
2           0  
3           0  
4           0


> **Q5.4 Short answer: [5 marks]**  
> In 2 to 3 sentences, explain why one hot encoding is suitable for `city_type`  
> and why ordinal encoding is suitable for `satisfaction_level`.

Write your answer here:

>  city_type has values that can't be justified by means of ranking one type of value greater than the other type, and the machine will be trained in the wrong manner.
>  Hence, for ensuring even distribution of priority, one hot encoding is suitable for city_type
>  On the other hand, the values in 'satisfaction level' makes sense to be ranked from high to low, hence ordinal ranking makes sense here.



---
## Part C - Module 3: Distances between users

For this small part we will work with vectors based on scaled numeric features.



### Q6. Euclidean and Manhattan distance

Build 2D vectors for user 0 and user 1 using:
- `income_std`  
- `daily_app_opens` (or its min max scaled version if you prefer)


In [53]:
# Q6.1: Build 2D vectors for first two users [We already write this ans]
vec_cols = ["monthly_income", "daily_app_opens"]

v1 = df.loc[0, vec_cols].values
v2 = df.loc[1, vec_cols].values

print("v1:", v1)
print("v2:", v2)

v1: [np.float64(3734.19) np.int64(48)]
v2: [np.float64(2594.19) np.int64(7)]


In [54]:
# Q6.2: Euclidean distance computation [5 marks]
dif = v1 - v2
eu = np.sqrt(np.dot(dif, dif))
print(round(eu, 3))

1140.737


In [57]:
# Q6.3: Manhattan distance computation [5 marks]
man = np.sum(np.abs(v1 - v2))
print(round(man, 3))

1181.0



> **Q6.4 Short answer: [5 marks]**  
> Which one is larger in your result, Euclidean or Manhattan distance  
> and why does that usually happen based on their formulas?

Write your answer here:

>  The Manhattan distance is larger than the euclidian distance.
>  This is because (a + b) > a^2 + b^2.



---
## Final Reflection [10 marks]

> In 4 to 6 sentences, describe how the three modules connect in this assignment.  
> Mention:
> - One idea from Module 1 or 2 that you used  
> - One idea from Module 3 that you used  
> - How these ideas together help you understand a dataset more deeply

Write your reflection here:

>  From Module 1, I used the concepts of mean and standard deviation.

>  From Module 2, I used the concepts of precision and recall.

>  From Module 3, I used the concepts of min-max scaling, ordinal encoding and manhattan distance

>  Through these concepts, I understood which values can cause the dataset to not be balanced, leading to the model to be biased. I also understood how a model's usefulness depends on how many values it can correctly predict. I also understood how different scaling and encoding techniques helps a model to be trained better.



## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।
